### DISCUSSION TITLE
How to specify system requirements for r-universe builds

### DISCUSSION THREAD

**Discussion by pepijn-devries** [r-universe-org/help#519] [Q&A] [answered]
Currently, I'm working on a package that uses static libraries:

https://github.com/pepijn-devries/openmpt

Therefore, it has system requirements ensuring that the libraries are available on the machine it runs on. I've included the requirements here:

https://github.com/rstudio/r-system-requirements

This will ensure that the rhub workflow on github will succefully build the package on Linux machines. However, when setting the package up for r-universe, the build fails (https://pepijn-devries.r-universe.dev/builds). This is because the system requirements are not installed.

Does anyone have know how to set this up correctly?

Many thanks for your help!

**Comment by jeroen** [comment]
It is explained here: https://docs.r-universe.dev/publish/debug-build.html#how-to-request-additional-third-party-system-requirements

If it is in r-system-requirements it should work. I'll have a look.

**Comment by jeroen** [comment]
Hmm if I run `remotes::system_requirements('ubuntu-22.04')` inside your package dir, the sysdep does not show up. Are you sure the regex matches?

**Reply by pepijn-devries** [reply]
> Hmm if I run `remotes::system_requirements('ubuntu-22.04')` inside your package dir, the sysdep does not show up. Are you sure the regex matches?

Thanks for having a look. It didn't at first, but after the discussion below, I had modified it such that it should match. The latest commit in the master branch seems to work together with https://github.com/pepijn-devries/openmpt/blob/master/.github/workflows/R-CMD-check.yaml

https://forum.posit.co/t/my-r-cmd-check-github-action-set-up-with-r-hub-fails-because-system-requirements-are-not-installed/194422/2

**Reply by pepijn-devries** [reply]
I got:

```r
# exiting from: remotes::system_requirements("ubuntu-20.04")
# [1] "apt-get install -y libavfilter-dev"
```

So it seems to pick up the `av` import from my description file, but not the system requirements. This is how the field is currently formatted in the DESCRIPTION file:

```
SystemRequirements:
    openmpt:
        libopenmpt-devel (rpm) or libopenmpt-dev (deb),
        portaudio-devel (rpm) or portaudio19-dev (deb).
```

And this is how the pattern in [openmpt.json](https://github.com/rstudio/r-system-requirements/blob/main/rules/openmpt.json) is specified:

```
  "patterns": ["\\bopenmpt\\b"],
```

If you have any advice on how to fix this, it's more than welcome!

**Reply by jeroen** [reply]
I'm going to try to switch to `pak` for installing sysreqs:

```
pak::pkg_sysreqs('.', sysreqs_platform='ubuntu-24.04')$install_script
```

**Reply by pepijn-devries** [reply]
Thanks, that does look promising...

**Reply by pepijn-devries** [reply]
It did help a bit. The source package is now built at r-universe, but the binary for Linux is not. This too is because of the missing system requirements

**Reply by jeroen** [reply]
Yes I know, need to fix this in several places...

**Reply by pepijn-devries** [reply]
Ah OK. I was too fast to test ;). Thanks again.

**Comment by jeroen** [comment, marked as answer]
It should work now, I switched the system dependency resolver to pak: https://github.com/r-universe-org/build-source/commit/15541772babad3fdab81dae270990f2f2b1b06bd

However this is still tentative because we had some troubles in the past with `pak` hanging, so if that happens again we may need to find another solution.

**Reply by pepijn-devries** [reply]
Thanks! In this case it seems to work for now!

**Comment by jeroen** [comment]
Ugh we may have to revert, `pak` is hanging again https://github.com/r-universe/adafede/actions/runs/11686437398/job/32542209386

**Reply by pepijn-devries** [reply]
OK. I thought learning to use static libraries in R would be fun :laughing:. At least I'm learning a lot :relieved:...

**Comment by jeroen** [comment]
Trying this workaround https://github.com/r-universe-org/build-source/commit/d2a1418bdfb5e8b8559a1a80abdb38590a2fa64a

**Comment by sebffischer** [comment]
I think I am also experiencing a related issue. 

I have my package anvil (https://github.com/r-xla/anvil) which depends on (https://github.com/r-xla/pjrt), which has system dependencies. 
I don't want to put it onto CRAN for now, so I put my r-xla r-universe into the `Additional_repositories` field: https://github.com/r-xla/anvil/blob/579f42a6d78d19e4b3f4459ac158070762225dff/DESCRIPTION#L35

In my github action CI, I also specify my r-universe as an extra-repository: https://github.com/r-xla/anvil/blob/579f42a6d78d19e4b3f4459ac158070762225dff/.github/workflows/R-CMD-check.yaml#L74, but for some reason, the github action workflow on ubuntu fails: https://github.com/r-xla/anvil/actions/runs/18902349676/job/53952342770 because the system dependency of anvil's dependence {pjrt} (libprotobuf etc.) is not installed properly.

**Reply by sebffischer** [reply]
The weird thing is that libprotobuf is detected and shown in the dependency lockfile, but protobuf-compiler isn't. But in the main branch of the repository, where I have anvil's dependencies in Remotes, it also detects protobuf-compiler: https://github.com/r-xla/anvil/actions/runs/18901830024/job/53950640382

**Reply by pepijn-devries** [reply]
What I can see is that `anvil` itself doesn't have system requirements, only dependencies. I think you should take care of system requirements for each individual package. Does `prjt` build without problems on R-Universe? In that case you could depend on that build in that universe. See https://docs.r-universe.dev/install/dependencies.html for details.

**Reply by sebffischer** [reply]
My problem is that when I install `anvil` in the CI, the system requirement of pjrt is not installed. I am taking care of system requirements in pjrt and pjrt builds successfully on r-universe on most platforms and where it does not build, the issues are not the system requirements.

**Reply by sebffischer** [reply]
Also the problem happens only when anvil installs the pjrt dependency from r-universe and not from GitHub, so I think the issue is related to r-universe

**Reply by jeroen** [reply]
I think I understand the issue. I'll try to fix it.

The problem is that R-universe automatically detects and normalizes the system dependencies, but it does not know that protobuf needs an extra package at build time. See the `_sysdeps ` field on https://r-xla.r-universe.dev/api/packages/pjrt

**Reply by sebffischer** [reply]
thanks!

**Reply by jeroen** [reply]
Added a workaround, could you test if your problem is solved now?

**Reply by sebffischer** [reply]
thanks, works like a charm!
